pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -214,64 +291,54 @@ func claimOwnerMatchesSetAndPod(logger klog.Logger, claim *v1.PersistentVolumeCl
 
 // updateClaimOwnerRefForSetAndPod updates the ownerRefs for the claim according to the deletion policy of
 // the StatefulSet. Returns true if the claim was changed and should be updated and false otherwise.
-func updateClaimOwnerRefForSetAndPod(logger klog.Logger, claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) bool {
-	needsUpdate := false
-	// Sometimes the version and kind are not set {pod,set}.TypeMeta. These are necessary for the ownerRef.
-	// This is the case both in real clusters and the unittests.
-	// TODO: there must be a better way to do this other than hardcoding the pod version?
-	updateMeta := func(tm *metav1.TypeMeta, kind string) {
-		if tm.APIVersion == "" {
-			if kind == "StatefulSet" {
-				tm.APIVersion = "apps/v1"
-			} else {
-				tm.APIVersion = "v1"
-			}
-		}
-		if tm.Kind == "" {
-			tm.Kind = kind
-		}
+func updateClaimOwnerRefForSetAndPod(logger klog.Logger, claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) {
+	originalRefs := claim.GetOwnerReferences()
+	newRefs := make([]metav1.OwnerReference, len(originalRefs))
+	copy(newRefs, originalRefs)
+
+	unexpectedController := hasUnexpectedController(claim, set, pod)
```

Comment:
What I found when refactoring this is that it gets very confusing & hard to maintain if we try to combine update logic and validation logic.

I think it will be a lot more understandable and reliable to have the update code just update and the validation code (isClaimOwnerUpToDate) just validate. Otherwise we can get into many weird corner cases and hard to reason about code.

Reply:
It's not necessary to call isClaimOwnerUpToDate first. After calling updateClaimOwnerRefForSetAndPod, usually the caller will make an api server update call which is potentially expensive, so calling isClaimOwnerUpToDate avoids that.

That seems to be the logic in stateful_set_control.go, eg see around line 442.